### DISCUSSION TITLE
passing args to included template

### DISCUSSION THREAD

**Discussion by stevegeo386** [sqlalchemy/mako#395] [Usage Questions] [answered]
A little cry for help.
I don't know if I misunderstand the documentation, if there is a bug, or I'm just trying to do something that mako cannot do. I've tried to search google for an answer, and just cannot seem to get my desired result.
the templates are in a subdirectory named 'makotemplates', as I wanted to keep the templates as close to my real work as possible.

In the template named parent.tmpl I have included via the <%include /> tag the other template called member.tmpl,  
<%include file="member.tmpl" args="mitem=member"/>
I am trying to pass a variable (a list of dictionary)  to the member.tmpl using the args tag, and expecting to see the 'mitem' variable in  member.tmpl but unfortunately I cannot 'see' the passed in arg ?

Have tried on multiple machines with different version of mako
Windows : py3 & mako 1.3.3
Windows : py3 & mako 1.2.4
Ubuntu 22.04 : py3 & mako 1.1.3
Ubuntu 20.04 : py2 & mako 1.1.0
Ubuntu 20.04 : py3 & mako 1.1.0
Debian 11: py3 & mako 1.2.4

all same response, 
Python311\site-packages\mako\runtime.py", line 793, in _include_file
    callable_(ctx, **kwargs)
  File "member_tmpl", line 26, in render_body
NameError: 'mitem' is not defined

here are my templates and an example python file

=== makotemplates/parent.tmpl =====
//----------------------------------------------------------------------
//  ${dd['summary']}
//
{
% for item in dd['charlie']:
<%
    if item['comment']:
        comment = "// {0}".format(item['comment'])
    else:
        comment = ''
    print("type(item.member) = {0}".format(type(item['member'])))
    member = item['member']
    print(type(member))
    print(member)
%>    ${comment}

 <%include file="member.tmpl" args="mitem=member"/>

% endfor
};

=== makotemplates/member.tmpl =====
    ## item should be of type StructMemberInit
<%
    print("type(mitem): {0}".format(type(mitem)))
    if mitem.MemberComment:
        comment = "// {0}".format(mitem.MemberComment)
    else:
        comment = ''
%>    ${mitem.MemberName} = ${mitem.MemberValue}, ${comment}

===== makoMWE2.py =====
from mako.lookup import TemplateLookup

# create a dict of data to send into the template

mystruct = {
    'bob' : [1,2,3,4],
    'charlie': [ 
        {'member' : {'value': 0x430, 'name': 'address', 'CS': 0x23}, 'comment': "c1"}, 
        {'member' : {'value': 11, 'name': 'offset', 'CS': 0x23}, 'comment': "c2"}, 
        {'member' : {'value': 4, 'name': 'size', 'CS': 0x23}, 'comment': "c3"}, 
        ],
    'comment' : "This is an example",
    'summary' : "This is a minimum working example",
}
   
mylookup = TemplateLookup(directories=['makotemplates'], strict_undefined=True)
mytemplate = mylookup.get_template( 'parent.tmpl' )
print(mytemplate.render( dd = mystruct ))

**Comment by zzzeek** [comment, marked as answer]
Hi -

First thing that will help, when posting code on github, [use triple backticks for formatting](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code).   I had to copy and paste your code examples and manually guess where the indentation was.

Next, in Python, when we have dictionary structures such as `mystruct = {"bob": "charlie"}`, we use indexed access like `mitem['ItemName']`, your example uses object syntax which does not work here based on the data you are passing.

Finally, to declare arguments in a mako template that must be passed, use [the page tag](https://docs.makotemplates.org/en/latest/syntax.html#page) to declare those variables.

your code can be formatted into a working demo as below

```py
from mako.lookup import TemplateLookup

lookup = TemplateLookup(strict_undefined=True)

lookup.put_string(
    "parent.tmpl",
    r"""
//----------------------------------------------------------------------
// ${dd['summary']}
//
{
% for item in dd['charlie']:
<%
if item['comment']:
    comment = "// {0}".format(item['comment'])
else:
    comment = ''

print("type(item.member) = {0}".format(type(item['member'])))
member = item['member']
print(type(member))
print(member)
%> ${comment}

<%include file="member.tmpl" args="mitem=member"/>

% endfor
};

""",
)

lookup.put_string(
    "member.tmpl",
    r"""
<%page args="mitem"/>
## item should be of type StructMemberInit
<%
    print("type(mitem): {0}".format(type(mitem)))
    if mitem['name']:
        comment = "// {0}".format(mitem['name'])
    else:
        comment = ''
%> ${mitem['CS']} = ${mitem['value']}, ${comment}

""",
)


mystruct = {
    "bob": [1, 2, 3, 4],
    "charlie": [
        {
            "member": {"value": 0x430, "name": "address", "CS": 0x23},
            "comment": "c1",
        },
        {
            "member": {"value": 11, "name": "offset", "CS": 0x23},
            "comment": "c2",
        },
        {"member": {"value": 4, "name": "size", "CS": 0x23}, "comment": "c3"},
    ],
    "comment": "This is an example",
    "summary": "This is a minimum working example",
}

mytemplate = lookup.get_template("parent.tmpl")
print(mytemplate.render(dd=mystruct))

```

**Reply by stevegeo386** [reply]
thank you zzzeek, first let me apologize for making it hard for you to help me, ..  I'll watch out for the formatting issue next time.

My problem was solved with the addition of <%page > tag, so thank you very much.

Also sorry for the confusion between dict and object dot notation, I was modifying my production code to a simple working example, and as my member template was not running I overlooked the fact I had not converted the variable names into dict index format.

Once again thank you for such a fast response.
